pkg/ingester/ingester_v2.go

```diff
@@ -82,6 +90,20 @@ func (i *Ingester) v2Push(ctx old_ctx.Context, req *client.WriteRequest) (*clien
 		return nil, err
 	}
 
+	// Ensure the ingester shutdown procedure hasn't started
+	i.userStatesMtx.RLock()
+
+	if i.stopped {
+		i.userStatesMtx.RUnlock()
+		return nil, fmt.Errorf("ingester stopping")
+	}
+
+	// Keep track of in-flight requests, in order to safely start blocks transfer
+	// (at shutdown) only once all in-flight write requests have completed
+	i.TSDBState.inflightWriteReqs.Add(1)
+	i.userStatesMtx.RUnlock()
```

Comment:
Do we need to do this with a lock? Can we not add it outside the lock?

Can this be simplified to:
```
i.userStatesMtx.RLock()
stopped := i.stopped
i.userStatesMtx.RUnLock()

if stopped {
}
```

Reply:
Unfortunately not because - in order to avoid any race condition - we also need to `i.TSDBState.inflightWriteReqs.Add(1)` within the lock.